types/operations/comparisons/PatternMatchingComparators.hpp

```diff
@@ -0,0 +1,240 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TYPES_OPERATIONS_COMPARISONS_PATTERN_MATCHING_COMPARATORS_HPP_
+#define QUICKSTEP_TYPES_OPERATIONS_COMPARISONS_PATTERN_MATCHING_COMPARATORS_HPP_
+
+#include <cstddef>
+#include <string>
+
+#include "catalog/CatalogTypedefs.hpp"
+#include "types/TypedValue.hpp"
+#include "types/operations/comparisons/Comparison.hpp"
+#include "types/port/strnlen.hpp"
+
+#include "re2/stringpiece.h"
+#include "re2/re2.h"
+
+namespace quickstep {
+
+class ColumnVector;
+class TupleIdSequence;
+class ValueAccessor;
+
+/** \addtogroup Types
+ *  @{
+ */
+
+/**
+ * @brief Base class for UncheckedComparators which does pattern matching.
+ **/
+template <bool is_like_pattern, bool is_negation,
+          bool left_nullable, bool right_nullable>
+class PatternMatchingUncheckedComparator : public UncheckedComparator {
+ public:
+  /**
+   * @brief Constructor.
+   **/
+  PatternMatchingUncheckedComparator(const std::size_t left_length,
+                                     const std::size_t right_length)
+    : left_length_(left_length),
+      right_length_(right_length) {
+  }
+
+  inline bool compareDataPtrs(const void *left,
+                              const void *right) const override {
+    return compareDataPtrsInl(left, right);
+  }
+
+  inline bool compareDataPtrsInl(const void *left,
+                                 const void *right) const {
+    if ((left_nullable && (left == nullptr))
+        || (right_nullable && (right == nullptr))) {
+      return false;
+    }
+
+    const char* text = static_cast<const char*>(left);
+    const re2::StringPiece text_piece(text, strnlen(text, left_length_));
```

Comment:
For all `strnlen` used in this file, either add `using std:: strnlen;` before the namespace, or add replace them with `std::strnlen`.